[scan] Autograd with partial gradient support

bohnstingl opened the pull request:
This PR introduces the Autograd feature for scan with partial gradient support. It is a combination of the already opened PRs: https://github.com/pytorch/pytorch/pull/135631 and https://github.com/bohnstingl/pytorch/pull/4

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @ydwu4

bohnstingl committed:
WIP: Integration of scan autograd

bohnstingl committed:
Introduced autograd for scan with partial gradients support

bohnstingl commented:
@pytorchbot label "topic: not user facing"

bohnstingl committed:
Fixed CI test issue with graph

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into scan_autograd22

ydwu4 reviewed (commented):
We're on the right direction. A big concern is readability, we should figure out a way to significantly reduce the complexity around all kinds of masking.

bohnstingl committed:
Integrated code review comments

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -49,6 +52,197 @@ def _extract_carry_and_out(flat_out: list[Any], num_carry: int):
+            # Get gradient masks of inits, xs and additional_inputs.
+            # These masks are used during the backward path to
+            # return gradients for inits with requires_grad=True and None for the others
+            init_mask = get_gradient_mask(init)
```
Done.

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -49,6 +52,197 @@ def _extract_carry_and_out(flat_out: list[Any], num_carry: int):
+def get_gradient_mask(tensor_list):
+    return [
+        True
+        if isinstance(v, torch.Tensor) and v is not None and v.requires_grad
```
True, the check for None is not necessary.

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -49,6 +52,197 @@ def _extract_carry_and_out(flat_out: list[Any], num_carry: int):
+                num_init,
+            )
+
+            fw_carry, fw_outputs = [pytree.tree_map(_from_fun, c) for c in carry], [
```
We don't. I have removed it.

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -49,6 +52,197 @@ def _extract_carry_and_out(flat_out: list[Any], num_carry: int):
+                raise RuntimeError(
+                    "Expect outputs produced by combine_fn to only contains tensors. "
+                    f"Got types {[type(out) for out in fw_outputs]}."
+                )
```
These checks are not necessary here and actually are copy/paste leftovers. I moved it to the front-end

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -49,6 +52,197 @@ def _extract_carry_and_out(flat_out: list[Any], num_carry: int):
+                    "The init and carries need to have the same require_grad structure! \
+                    E.g., check for `with torch.no_gard` statements in the combine_fn."
+                )
+
```
I moved this check

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -49,6 +52,197 @@ def _extract_carry_and_out(flat_out: list[Any], num_carry: int):
+                combine_fn,
+                False,
+                (
+                    *fw_init,
```
No, not necessarily. The all `fw_init`, `fw_xs` and `fw_additional_inputs` are required even if they don't require a gradient. This is because the joint_graph wouldn't work otherwise. In the revised version we return torch.zeros_like() for the elements that don't require gradients.

Moreover, is there anything from https://github.com/pytorch/pytorch/pull/142518 to consider here, or is there anything blocking CUDA graph support? I hope I replaced all the Nones with torch.zeros_like(), but I may have overlook something.

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -49,6 +52,197 @@ def _extract_carry_and_out(flat_out: list[Any], num_carry: int):
+                carried_g_additional_input = args[:num_additional_inputs]
+
+                g_c, g_xs = _extract_carry_and_out(
+                    joint_graph(*args[num_additional_inputs:]), num_init
```
I tried to simplify this in a new version. The new flow is as follows:

1. Create the forward and the joint graph of the `combine_fn`
2. Retrace the wrapper of the forward graph that returns carries from all iterations, not just from the last iteration
3. Obtain the gradients from the joint_graph and compute the gradient masks.
4. Retrace the wrapper for the joint_graph using the masks.

Is this more clear? I have also added some more comments.

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -272,19 +468,18 @@ def _scan(init, xs):
-                for i, e in enumerate(dummy_out)
-            ]
-        )
+        outs = [
+            torch.zeros(
+                [num_elems] + list(e.size()),
+                dtype=e.dtype,
+                device=e.device,
+            )
+            for i, e in enumerate(dummy_out)
+        ]
+        idxs = [
+            torch.ones_like(e, dtype=torch.int64).unsqueeze(0)
+            for i, e in enumerate(dummy_out)
+        ]
```
I'd think so. This is for storing the temporary outputs of scan

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into scan_autograd22

bohnstingl committed:
Fixed type annotation

bohnstingl committed:
Removed reverse flag from backend and implemented reverse with torch.flip in the frontend

bohnstingl committed:
Fix of graph in testcase

bohnstingl committed:
Merge branch 'scan_flip_reverse' of github.com:bohnstingl/pytorch into scan_autograd22

bohnstingl committed:
Integrated new reverse handling into scan autograd

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into scan_autograd22

bohnstingl committed:
Fixed issues with testcases and with combine_fn now returning pytrees

ydwu4 committed:
[cond] don't trace fw and bw graph in autograd key

[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] don't trace fw and bw graph in autograd key"




[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] don't trace fw and bw graph in autograd key"




[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] don't trace fw and bw graph in autograd key"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] don't trace fw and bw graph in autograd key"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into scan_autograd22

bohnstingl committed:
Merge branch 'gh/ydwu4/222/head' of github.com:pytorch/pytorch into scan_autograd22_rework

ydwu4 committed:
Update base for Update on "[cond] don't trace fw and bw graph in autograd key"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] don't trace fw and bw graph in autograd key"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] don't trace fw and bw graph in autograd key"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 committed:
Update base for Update on "[cond] don't trace fw and bw graph in autograd key"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] don't trace fw and bw graph in autograd key"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 committed:
Update base for Update on "[cond] don't trace fw and bw graph in autograd key"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bohnstingl committed:
WIP: scan autograd

bohnstingl committed:
Merge branch 'gh/ydwu4/222/head' of github.com:pytorch/pytorch into scan_autograd22_rework

ydwu4 committed:
Update base for Update on "[cond] don't trace fw and bw graph in autograd key"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] don't trace fw and bw graph in autograd key"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bohnstingl committed:
Merge branch 'gh/ydwu4/222/head' of github.com:pytorch/pytorch into scan_autograd22_rework

bohnstingl committed:
Working function of scan autograd with new infrastructure

bohnstingl committed:
Removed unnecessary code and other cleanups

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into scan_autograd22

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into scan_autograd22

bohnstingl committed:
Fixed tracing issue causing missing fake meta values

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into scan_autograd22

bohnstingl committed:
Fixed CI test

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into scan_autograd22

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into scan_autograd22

bohnstingl committed:
Fixed testcase

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +387,309 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+        # 3.) Materialize the combine_fn_bw
+        # TODO: we need to materialize the bw graphs because dynamo is unable to
+        # trace through the joint function when torch.compile torch.autograd.grad.
+        combine_fn_bw_gm = materialize_as_graph(
+            ctx._combine_fn_bw,
+            (
+                *init,
+                *xs_slice,
+                *additional_inputs,
+                *[first_slice_copy(c) for c in carries],
+                *[first_slice_copy(o) for o in outs],
+            ),
+            ctx._fw_include_key_set,
+            ctx._fw_exclude_key_set,
+            force_enable_grad=True,
+        )
```
We could nest the python function and materialize the graph in the end. so materializing the function once would be good enough.

bohnstingl replied:
Done

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +387,309 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+        def wrapper_bwd_combine_fn(*args):
+            carried_g_additional_input = args[: ctx._num_additional_inputs]
+
+            # Adjust the order of the args.
+            # We get tangents gradients + inputs, but the joint function expects inputs + gradients
+            shuffled_args = [
+                *args[: ctx._num_additional_inputs],
+                *args[
+                    -(
+                        ctx._num_leaves_init
+                        + ctx._num_leaves_xs
+                        + ctx._num_additional_inputs
+                    ) :
+                ],
+                *args[
+                    ctx._num_additional_inputs : -(
+                        ctx._num_leaves_init
+                        + ctx._num_leaves_xs
+                        + ctx._num_additional_inputs
+                    )
+                ],
+            ]
```
we could partition args and give them a meaningful variable name so that the re-shuflling logic becomes more sensible (which becomes re-ornazing the variables)?

bohnstingl replied:
I removed the slicing and made the parts more sensible and simpler to understand

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +387,309 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+            ]
+
+            g_c, g_xs = _extract_carry_and_out(
+                combine_fn_bw_gm(*shuffled_args[ctx._num_additional_inputs :]),
```
can just call ctx._combine_fn_bw instead of materialzing combine_fn_bw_gm.

bohnstingl replied:
Done

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +387,309 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+                num_leaves_init,
+            )
+
+            current_g_additional_inputs = g_xs[len(g_xs) - ctx._num_additional_inputs :]
```
I don't like slicing because it will lose track of the semantic meaning, can we name each slice properly upfront?

bohnstingl replied:
I should have improved this

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +387,309 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+
+            return [*new_g_additional_inputs, *g_c, *g_xs]
+
+        ctx._combine_fn_bw_wrapped = _maybe_reenter_make_fx(wrapper_bwd_combine_fn)(
```
can call materialize_graph here.

bohnstingl replied:
Done

ydwu4 reviewed (commented):
Overall looks good but I have serious concerns for code readability.

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +387,309 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+
+        Example::
+
+            The ``combine_fn`` f(.,.), used in the forward function, is the operator used during the scan. For example
```
The doc is hard to read.

One way to improve the doc is to present the pseuodo code for its implementation so that after reading it, we know the overall plan.

bohnstingl replied:
I improved the doc including with a pseudocode

ydwu4 reviewed (commented):
To summarize:
1. I only want to call materialize_graph once in backward. No _reenter_make_fx or anything that turns a function into graph except materialize_graph.
2. I want a clearer contract on semantic meaning the inputs to the subgraph/fn (each slicing op need to have a clear definition). The input and output of each wrapper fn have semantic meanings.
3. documenting the **rationale** of anything that's non-trivial. Many of the code comments don't make senese because they're just explaining the **what** instead of **why**. e.g. for line 643-678 what's the overall design and why we need prepare initiali gradients and carries.

bohnstingl committed:
Integrated review feedback; before code cleanup

bohnstingl committed:
After cleanup and lintrunner

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +387,309 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+
+                combine_fn_wrapped = _maybe_reenter_make_fx(wrapper_fwd_combine_fn)(
+                    *fw_init, *fw_xs, *fw_additional_inputs
+                )
```
Done

bohnstingl commented:
@ydwu4 thank you for your review. You raised good points, indeed. I tried to address them in an update to this PR. In particular:

1. I have changed the backward such that the `materialize_graph` is only called once during the backward and not other function that creates a graph is used.
2. I removed most of the slicing and assigned meaningful names to the individual parts such that it becomes clearer what their semantics is and how they are being used
3. I enhanced the documentation of the backward quite a bit and now every non-trivial line / code segment should be documented. This includes adding a pseudocode in the description at the beginning of the backward.

What do you think about this version?

bohnstingl committed:
Fixes

bohnstingl committed:
Further cleanup and restructuring

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +394,292 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+
+        # 4.) Create the BW wrapper
+        def combine_fn_bw_wrapped(*args):
+            # Dissect arguments
+            (
+                carried_g_additional_input,
+                carries_g,
+                outs_g,
+                init,
+                xs,
+                additional_inputs,
+            ) = split_into_chunks(
+                args,
+                [
+                    num_additional_inputs,
+                    num_leaves_init,
+                    num_leaves_ys,
+                    num_leaves_init,
+                    num_leaves_xs,
+                    num_additional_inputs,
+                ],
+            )
+
+            # Adjust the order of the variables in args.
+            # We get tangents gradients + inputs, but the joint function expects inputs + gradients
+            rearranged_args = [*init, *xs, *additional_inputs, *carries_g, *outs_g]
+
```
nit: Can just concat?

bohnstingl replied:
I am not quite sure where and how concat would help here?

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +394,292 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+                    )
+                ],
+                *[first_slice_copy(c) for c in fw_carries],
+                *[first_slice_copy(o) for o in fw_ys],
```
what's the difference between first_slice_copy_with_grad and first_clice_copy? Why we need two? Can we move the logic of computing fw_xs_slice here?

Can have a function "construct_args_single_step_bw" that documents why we set args to be this order, why we need slicing, why we need masking.

bohnstingl replied:
The difference between the two is that `first_slice_copy` does not preserve the requires_grad flag. This flag though is essential for the `materialize_as_graph` function to work properly.

I created a dedicated function `construct_args_single_step_bw` that describes the arguments for the `materialize_as_graph` quite verbosely.

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +394,292 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+
+            # Adjust the order of the variables in args.
+            # We get tangents gradients + inputs, but the joint function expects inputs + gradients
+            rearranged_args = [*init, *xs, *additional_inputs, *carries_g, *outs_g]
```
Can we avoid the split and reorder? since we have control over how the fn is called.

bohnstingl replied:
No, I don't think this can be easily done. the issue is that the `ctx._combine_fn_bw` as created through `create_bw_fn` does expect the order `primals` followed by `tangents`. However, the `scan` in the backward uses init and xs to provide those values, but init are the gradients and xs are the primals in that case, so `scan` provides `tangents` followed by `primals`. This is why we need to rearrange the inputs here. I wouldn't really know how to change that in the current code. I just see two alternatives:

1. Create a wrapper around `ctx._combine_fn_bw` that does the reordering, so then we don't need to do it in the backward in this awkward way.
2. Create a specific `create_bw_fn` for scan that has `tangents` followed by `primals`.

I opted for the first option and created a wrapper that takes care of the ordering. Does this improve things?

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -33,6 +36,15 @@
 aten = torch._ops.ops.aten
 
 
+def split_into_chunks(iterable: Iterable[Any], chunk_sizes: list[int]) -> list[Any]:
```
can we assert sum(chunk_sizes) == len(iterable)?

bohnstingl replied:
Done.

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -134,6 +175,14 @@ def _validate_input(cfn, lxs, linit, d, r):
         for x in lxs:
             if not isinstance(x, torch.Tensor):
                 raise RuntimeError(f"All xs leaves must be a Tensor but got {x}")
+        if any(x.ndim <= d for x in lxs):
```
Necessary for this PR?

bohnstingl replied:
Yes, some of the reworked tests require and would else report a "non-user friendly" error message to the user.

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +394,292 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+        )
+
+        # Decompose the flat_grads into g_c_T, g_ys
+        g_c_T, g_ys = split_into_chunks(flat_grads, [num_leaves_init, num_leaves_ys])
```
what's g_c_T Can change its name to g_last_carry?

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +394,292 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+        # ``wrapper_bwd_combine_fn`` and thus need a zero-initialized starting point
+        initial_g_additional_inputs = initialize_g_additional_inputs(additional_inputs)
+
+        # Prepend the inits to the carries.
+        # This is needed, because when computing the gradients, the last carry is not needed
+        # but the first carry, the init, is required.
+        bw_carries = prepend_init_to_carries(fw_init, fw_carries)
+
+        # Prepare the xs for the backward scan.
+        bwd_xs = [*g_ys, *bw_carries, *fw_xs]
+
+        # The flipping of the ``bwd_xs`` is necessary because the scan_op in the backward is always performed in reverse
+        bwd_xs = [torch.flip(elem, [0]) for elem in bwd_xs]
+
+        # Prepare the bwd_init
+        bwd_init = [*initial_g_additional_inputs, *g_c_T]
+
+        # 6.) Perform the backwrad scan:
+        # The ``combine_fn_bw_wrapped`` receives the
+        # initial_g_additional_inputs and the last carry as the ``bwd_init`` and the
+        # gradients of the outputs (g_ys), as well as the fw_carries and the fw_xs of the forward as the ``bwd_xs``
+        gradients = scan_op(
+            combine_fn_bw_wrapped_gm,
+            bwd_init,
+            bwd_xs,
+            additional_inputs,
+        )
```
I like the implementation here. It makes the backward scan's semantic super clear. I think we can put this part in the NOTE and explain why after we construct backward in this way, we'll be able to get grad_init and grad_xs and grad_additional_inputs.

bohnstingl replied:
I enhance the description quite a bit with details about the backward scan procedure and how in the end we get `g_init`, `g_xs` and `g_additional_inputs`.

ydwu4 commented on torch/_higher_order_ops/utils.py:
```diff
@@ -437,7 +437,7 @@ def fw_with_masks(*args):
         # Instead of using the original fn, we set the output of original
         # fn to all require grad. This is consistent with the behavior
-        # of autograd.Function, where if any one of the inputs requires grad
+        # of autograd. Function, where if any one of the inputs requires grad
```
accidental change?

bohnstingl replied:
Yeah, not really though. I spotted the typo and wanted to correct it. But I see this may not be wanted and thus I reverted the change and we can have it in another PR.

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -265,25 +314,17 @@ def store_out_in_outs(out, ind):
     return scans
 
 
-# We also do a clone with contiguous_format. This is to be consistent with
```
necessary for this PR?

bohnstingl replied:
Well, I just moved the function to have all of the helper function at the top of the file. If this is not wanted, I can put it back to the old place. I didn't change the code of that function.

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -2182,25 +2269,21 @@ def test_scan_compile_cnt(self, reverse, device):
-    def test_scan_init_scanned_0(self, compile_mode):
-        scan_fct = compile_mode_helper(scan, compile_mode)
-
+    @parametrize("device", [torch.device("cpu"), torch.device("cuda")])
```
What's going on with this test.

bohnstingl replied:
I should have corrected this test

ydwu4 replied:
not need to parametrize device i feel

bohnstingl replied:
I removed the parametrization

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -2209,28 +2292,16 @@ def test_scan_init_scanned_0(self, compile_mode):
-        scan_fct = compile_mode_helper(scan, compile_mode)
-
-        x = torch.randn(3, 1, 2)
+    @parametrize("reverse", [False, True])
```
what's going on with this test

bohnstingl replied:
I should have corrected this test

ydwu4 replied:
not need to parametrize device i feel

bohnstingl replied:
I removed the parametrization

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -2442,9 +2518,58 @@ def add_scalar_carry_sliced_out(x: torch.Tensor, y: torch.Tensor):
+    @requires_cuda
+    @parametrize("reverse", [False, True])
+    @parametrize("device", [torch.device("cpu"), torch.device("cuda")])
+    def test_scan_carry_wrong_pytree(self, reverse, device):
```
This change is not related to autograd, right? Can remove the test?

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -2652,45 +2796,525 @@ def forward(self, child: "f32[1, 10, 2]", child_1: "f32[1, 10, 2]", child_2: "f3
+    @parametrize("reverse", [False, True])
+    @parametrize("device", [torch.device("cpu"), torch.device("cuda")])
+    @parametrize("autograd", [False, True])
+    def test_scan_closure_RNN_parameters_as_inputs(self, reverse, device, autograd):
```
This test is not related to augrad i think?

bohnstingl replied:
Test removed.

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -2652,45 +2796,525 @@ def forward(self, child: "f32[1, 10, 2]", child_1: "f32[1, 10, 2]", child_2: "f3
+    ):
+        dim = 1
+        x = torch.randn(3, 10, 7, device=device, requires_grad=autograd)
+        h1 = torch.randn(3, 7, device=device, requires_grad=autograd)
```
does this test do what it supposed to do?

bohnstingl replied:
Good catch! I corrected this testcase.

ydwu4 replied:
"init_carries_unequal_grad" meaning init and carry have different require_grad? or something else?

bohnstingl replied:
Yes, precisely. In this test, there is a `no_grad` statement contained in the `combine_fn`. init starts out with `requires_grad=True` and then the carry however comes back with `requires_grad=False`. This is a special case of partial gradients and this is what this test is supposed to test.

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -2652,45 +2796,525 @@ def forward(self, child: "f32[1, 10, 2]", child_1: "f32[1, 10, 2]", child_2: "f3
+    @requires_cuda
+    @parametrize("reverse", [False, True])
+    @parametrize("device", [torch.device("cpu"), torch.device("cuda")])
+    @parametrize("autograd", [True])
```
autograd [True, False]? or remove the parameterization

bohnstingl replied:
I corrected this testcase.

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -2652,45 +2796,525 @@ def forward(self, child: "f32[1, 10, 2]", child_1: "f32[1, 10, 2]", child_2: "f3
 
     @skipIfNoDynamoSupport
-    def test_scan_simple_graph_wrong_dtype(self):
+    def test_scan_simple_graph_no_carry(self):
```
Not related to autograd?

bohnstingl replied:
Testcase removed.

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -2652,45 +2796,525 @@ def forward(self, child: "f32[1, 10, 2]", child_1: "f32[1, 10, 2]", child_2: "f3
+            )
+
+    @skipIfNoDynamoSupport
+    def test_scan_simple_graph_wrong_carry(self):
```
not related to autograd

bohnstingl replied:
Testcase removed

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -2705,7 +3329,7 @@ def f(fct, init, xs):
             "Expected init and carry to have same metadata.*",
         ):
-            f(add_wrong_dtype, init, x)
+            make_fx(f, tracing_mode="symbolic")(add_wrong_dtype, init, x)
```
why?

bohnstingl replied:
I reverted that change to the old version. It was an experimental change.

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -2761,6 +3386,24 @@ def forward(self, L_init_ : torch.Tensor, L_xs_ : torch.Tensor):
         )
 
+    @skipIfNoDynamoSupport
+    def test_scan_simple_graph_wrong_dtype(self):
```
not related to autograd.

bohnstingl replied:
Testcase removed

ydwu4 reviewed (commented):
Looks much better. Really appreciate for the efforts. We need some more polishing to get the PR ready.

bohnstingl committed:
WIP: integration of code review

bohnstingl committed:
Added extensive documentation and lintrunner corrections

bohnstingl committed:
Revert change to cond.py

bohnstingl committed:
Removed unrelated testcases and fixed lint issues

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +394,292 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+                next_carry, y = combine_fn(x, y)
+                return next_carry, (next_carry, y)
+
+            With the function defined above, and the forward carries and the outputs computed in step 3.) above:
```
Done, I moved the description and added a note for what happens to the gradients of elements that have `requires_grad=False`

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +394,292 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+
+            the backward of scan can be computed as:
+
+            4.) Create the wrapper of the ``combine_fn_bw`` that takes care of accumulating the gradients for the additional inputs
```
I went through the description and tried to update the names and reference them more clearly to the code segments. I also added a sentence on the additional inputs

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +394,292 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+            6.) Perform the backwrad scan as
+            g_additional_inputs, g_init, g_xs = scan_op(combine_fn_bw_wrapped, bw_init, bw_xs), where
+            bw_init is the last carry from the forward and
+            bw_xs is the combination of the upstream gradients g_ys, the forward carries prepended with the fw_init and the fw_xs
```
I added the definition of the `bw_init` and the `bw_xs`.

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +394,292 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+            4.) Create the wrapper of the ``combine_fn_bw`` that takes care of accumulating the gradients for the additional inputs
+            5.) Materialize the ``combine_fn_bw_wrapped``
+            6.) Perform the backwrad scan as
+            g_additional_inputs, g_init, g_xs = scan_op(combine_fn_bw_wrapped, bw_init, bw_xs), where
```
I added a sentence to clarify where the `g_last_carry` and `g_ys` come from. I hope this is sufficient?

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +394,292 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+        )
+
+        # First_slice_copy does not keep the original requires_grad flag,
+        # but we need it here in order to compute the correcte gradients
```
Yes, this is correct. We only need to have the gradient flag preserved correctly for the `materialize_as_graph` function. Thus, we only need the `first_slice_copy_with_grad` function once.

bohnstingl commented:
@ydwu4 I tried to incorporate all your requests and I think the PR would be ready for another round.

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +406,380 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+    bw_xs is the combination of the upstream gradients g_ys, the forward carries prepended with the fw_init and the fw_xs,
+    i.e., bwd_xs = [*g_ys, *bw_carries, *fw_xs], with bw_carries = concat([fw_init, fw_carries[:-1]]).
+
+    For ease of understanding the procedure of the gradient calculation is as follows:
+    One starts from the last step with the init being the upstream gradient of initial_g_additional_inputs (all zeros)
+    and g_last_carry.
+    Then, in the first scan step, we compute the gradients g_c_T, g_xs_T and g_addititional_inputs_T with
+    g_c_(T-1), g_xs_T and g_addititional_inputs_T = bw_fn(c_(T-1), xs_T, additional_inputs, g_last_carry, g_ys_T)
+    We then accumulate g_addititional_inputs_T with the g_addititional_inputs_T and thus use as the init for the next step
+    initial_g_additional_inputs + g_addititional_inputs_T, g_c_(T-1). Then, in the next step we compute the gradients for T-1, i.e.,
+    g_c_(T-2), g_xs_(T-1) and g_addititional_inputs_(T-1) = bw_fn(c_(T-2), xs_(T-1), additional_inputs, g_c_(T-1), g_ys_(T-1)).
+    We again accumulate the g_addititional_inputs_(T-1) and use it together with the g_c_(T-2) as the new init for the next step.
+    This procedure continues until we arrive at the first step, i.e.,
+    0, g_xs_0 and g_addititional_inputs_0 = bw_fn(init, xs_0, additional_inputs, g_init, g_ys_0).
+    Through this procedure we end up with the
+    gradients for the init -> g_init,
+    the gradients for the xs -> g_xs and
+    the gradients for the additional_inputs -> g_additional_inputs.
```
This should be put into earlier of this documentation. See my previous comment on how to organize the overall design.

bohnstingl replied:
I revised the document quite a bit. Let me know whether it is more clear now.

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +406,380 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+
+    As a last step, we mask the g_additional_inputs with Nones at places where the additional inputs are not tensors.
+
+    Note: Because we start with the last step T and gradually progress backward in time to the first step,
+    the elements of bw_xs are flipped along the scan dimension
```
this will be made clear with the overall design.

bohnstingl replied:
I removed that piece

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +406,380 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+    Note: Because we start with the last step T and gradually progress backward in time to the first step,
+    the elements of bw_xs are flipped along the scan dimension
+
+    Note: g_last_carry and g_ys are provided through the upstream autograd infrastructure
+
```
More specifically, autograd.Function.backward's input

bohnstingl replied:
I more it more specific.

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +406,380 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+
+    Note: g_last_carry and g_ys are provided through the upstream autograd infrastructure
+
+    Note: The scan_op in the backward needs to operate always reverse over time, i.e., starting from the last time step and
+    moving to the first. Therefore, the bwd_xs and the resulting gradients g_xs need to be flipped
```
redundant?

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +406,380 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+    Note: The scan_op in the backward needs to operate always reverse over time, i.e., starting from the last time step and
+    moving to the first. Therefore, the bwd_xs and the resulting gradients g_xs need to be flipped
+
+    Note: If any element of init, of xs or of the outputs does not require gradients, i.e., requires_grad=False,
+    there will be still gradients returned for those elements.
+    However, those gradients will be a tensor of the same shape as the element, but the gradient will be filled with zeros.
```
a seperate NOTE [scan partial grad handling]

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +406,380 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+            ctx._num_leaves_ys = len(outs)
+
+            # 3.) Prepare the backward graph
+            def combine_fn_bw(fn, ops):
+                n_primals = len(ops)
+
+                bw_fn = create_bw_fn(
+                    fn,
+                    ops,
+                )
+
+                def bw_fn_args_reordered(*args_and_grad_outs):
+                    # Change the order of the primals and the tangents
+                    # This is required because create_bw_fn creates the backward function such
+                    # that it requires primals followed by tangents, while the backward scan provides
+                    # tangents followed by primals. Therefore, this wrapper simply corrects this order mismatch.
+                    tangents = args_and_grad_outs[: len(args_and_grad_outs) - n_primals]
+                    primals = args_and_grad_outs[len(args_and_grad_outs) - n_primals :]
+                    return bw_fn(*primals, *tangents)
+
+                return bw_fn_args_reordered
```
why do we need the reorder here? we could do the reordering in construct_args_single_step_bw, no?

bohnstingl replied:
As we discussed offline, the reordering is needed. I have put it back into `combine_fn_bw_grad_accumulation`.

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +406,380 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+        # trace through the joint function when torch.compile torch.autograd.grad.
+        combine_fn_bw_grad_accumulation_gm = materialize_as_graph(
+            combine_fn_bw_grad_accumulation,
+            (*args_single_step_bw,),
```
nit: (*args_single_step_bw,) -> args_single_step_bw

bohnstingl replied:
Done

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +406,380 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+        # The flipping back along the scan dimension is required to get the gradients in the right order for ``xs``
+        g_xs = [torch.flip(elem, [0]) for elem in g_xs]
+
+        # The gradients for additional inputs that are not tensors are replaced with None.
```
Mention the partial grad handling notes.

bohnstingl replied:
I referred to that note.

ydwu4 reviewed (commented):
Mainly have some concerns about the organization of the documentation.

bohnstingl committed:
Integrated review changes for documentation and test cases

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -45,7 +47,58 @@ def wrap_combine_fn_flat(
+
+def mask_list(
+    mask: list[bool], inp: list[Any], other: Optional[list[Any]] = None
+) -> list[Any]:
```
Done.

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -45,7 +47,58 @@ def wrap_combine_fn_flat(
+        return [i for m, i in zip(mask, inp) if m]
+
+
+def first_slice_copy_with_grad(li):
```
Done.

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +406,380 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+        next_carry, y = combine_fn(x, y)
+        return next_carry, (next_carry, y)
+
+    NOTE: [scan forward implementation]
```
I removed the sub-notes and only kept `NOTE: [scan autograd implementation]`

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +406,380 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+    With the function defined as above, the forward output of scan works as follows:
+
+    1.) Prepare the forward graph wrapper ``combine_fn_with_carry_checkpoint``:
+    As mentioned above, we need the carries from all steps later for the backward path and thus we first prepare
```
Done

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +406,380 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+    the wrapper ``combine_fn_with_carry_checkpoint``, which produces as outputs all carries, the last carry and all outputs.
+
+    2.) Compute the all carries, the last carry and all outputs using ``combine_fn_with_carry_checkpoint``:
+    Next we utilize the ``combine_fn_with_carry_checkpoint`` and compute all carries, the last carry and all outputs, i.e.,
```
I reversed the order.

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +406,380 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+
+    The bw_fn requires the primals (operands) followed by the tangents (upstream gradients) from a single step
+    and produces the gradients of that step, i.e.,
+    g_c_(T-1), g_xs_T, g_additional_input_T = bw_fn(c_(T-1), xs_T, additional_inputs, g_last_carry, g_ys_T).
```
Yes, this is what I had before. I changed it back.

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +406,380 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+
+    Because we utilize the ``bw_fn`` in combination with scan during the backward function which provides
+    the arguments in the order tangents followed by primals, we need to wrap ``bw_fn`` into ``bw_fn_args_reordered``
+    which corrects the order from tangents followed by primals to the expected order of primals followed by tangents.
```
Okay, I removed this piece from the document, because we already have a description with inlined comments.

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +406,380 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+    the arguments in the order tangents followed by primals, we need to wrap ``bw_fn`` into ``bw_fn_args_reordered``
+    which corrects the order from tangents followed by primals to the expected order of primals followed by tangents.
+
+    NOTE: [scan backward implementation]
```
Removed.

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +414,393 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+        ...
+        g_init, g_xs_1 = combine_fn_bw(c_0, xs_1, g_c_0, g_ys_1)
+        0     , g_xs_0 = combine_fn_bw(init, xs_0, g_init, g_ys_0)
+
```
adding explaination "where combine_fn_bw takes fw inputs of step T (i.e. c_(T-1), xs_T), gradient of outputs of step T (i.e. g_c_T, g_ys_T) and returns the gradient of fw inputs (i.e. g_c_(T-1), g_xs_T)"

bohnstingl replied:
Explanation added.

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +414,393 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+    2.) Compute all carries, the last carry and all outputs using ``combine_fn_with_carry_checkpoint``:
+    c_T, (carries, ys) = scan_op(combine_fn_with_carry_checkpoint, init, xs, additional_inputs),
+    Where c_T (last carry) and ys (all outputs) are the original results of scan with the ``combine_fn``.
+    However, carries are checkpointed carries from all steps.
```
Can emphasize that only c_T and ys will be returned as forward outputs. carries will be saved for use in backward.

bohnstingl replied:
I added, a sentence to clarify.

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +414,393 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+    3.) Prepare the backward graph:
+    We prepare the backward graph to be used in the backward function.
+    We utilize ``create_bw_fn`` to generate the joint function, i.e.,
+    bw_fn = create_bw_fn(combine_fn, operands), where operands = [init, xs, additional_inputs]
```
Can we move the create_bw_fn logic into backward and check save the fw_fn. i think that would be more intuitive. Otherwise, people might be curious why it's necessary we need to call create_bw_fn in forward (the answer to that would probably be we don't need to).

bohnstingl replied:
True. I moved the create_bw_fn logic into the backward.

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +414,393 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+        inputs_bw_fn = args[num_additional_inputs:]
+        g_c_(t-1), g_xs_t, g_additional_input_t = bw_fn(*inputs_bw_fn)
+        new_g_additional_inputs = carried_g_additional_input + g_additional_input_t
+        return [*new_g_additional_inputs, *g_c_t, *g_xs_t]
```
emphasize that new_g_additional_inputs and g_c_t is encoded as part of carry while *g_xs_t are encoded as y.

bohnstingl replied:
I added two sentences to clarify.

ydwu4 reviewed (approved):
Looks good! Left some comments for more clarity of the doc. Also need to fix the test failure.

bohnstingl committed:
Added some more comments

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -353,9 +406,380 @@ def scan_op_dense(combine_fn, init, xs, additional_inputs):
+                    # tangents followed by primals. Therefore, this wrapper simply corrects this order mismatch.
+                    tangents = args_and_grad_outs[: len(args_and_grad_outs) - n_primals]
+                    primals = args_and_grad_outs[len(args_and_grad_outs) - n_primals :]
+                    return bw_fn(*primals, *tangents)
```
No, like mentioned here the backward scan invokes the wrapper with `tangents` followed by `primals` as `args_and_grad_outs` and thus we need to reorder here. This here is the reordering from `[tangents,primals]` to `[primals,tangents]`. This comment somehow refers to an outdated code section. I've changed the nomenclature in the current version to be more clear.

bohnstingl committed:
Introduced materialize_as_graph for combine_fn and for combine_bw_fn

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into scan_autograd22

bohnstingl committed:
Added bypass for autograd DispatchKey if no gradient for either the init, the xs or the additional_inputs is required

ydwu4 commented:
@pytorchbot merge

ydwu4 commented:
@pytorchbot merge -i
